core/shared/src/main/scala/zio/Config.scala

```diff
@@ -0,0 +1,289 @@
+/*
+ * Copyright 2022 John A. De Goes and the ZIO Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package zio
+
+import scala.util.control.{NonFatal, NoStackTrace}
+
+/**
+ * A [[zio.Config]] describes the structure of some configuration data.
+ */
+sealed trait Config[+A] { self =>
+
+  /**
+   * Returns a new config that is the composition of this config and the
+   * specified config.
+   */
+  def ++[B](that: Config[B])(implicit zippable: Zippable[A, B]): Config[zippable.Out] =
+    Config.Zipped[A, B, zippable.Out](self, that, zippable)
+
+  /**
+   * Returns a config whose structure is preferentially described by this
+   * config, but which falls back to the specified config if there is an issue
+   * reading from this config.
+   */
+  def ||[A1 >: A](that: Config[A1]): Config[A1] = Config.Fallback(self, that)
+
+  /**
+   * Adds a description to this configuration, which is intended for humans.
+   */
+  def ??(label: String): Config[A] = Config.Described(self, label)
+
+  /**
+   * Returns a new config whose structure is the same as this one, but which
+   * produces a different Scala value, constructed using the specified function.
+   */
+  def map[B](f: A => B): Config[B] = self.mapOrFail(a => Right(f(a)))
+
+  /**
+   * Returns a new config whose structure is the samea as this one, but which
+   * may produce a different Scala value, constructed using the specified
+   * fallible function.
+   */
+  def mapOrFail[B](f: A => Either[Config.Error, B]): Config[B] = Config.MapOrFail(self, f)
+
+  /**
+   * Returns a new config whose structure is the same as this one, but which may
+   * produce a different Scala value, constructed using the specified function,
+   * which may throw exceptions that will be translated into validation errors.
+   */
+  def mapOrThrow[B](f: A => B): Config[B] =
```

Comment:
Wouldn't `mapAndCatch` be a better name? There's no throwing taking place, the point of the function is that it _catches_ exceptions, right?
Or `mapOrError` or `mapOrInvalidData` or `mapOrInvalid`?

Reply:
The function `f` maps from `A` to `B` or it throws, similar to `mapOrFail`, which is widely used in ZIO.

Reply:
Just to make myself clear, what I found confusing is that the name suggests that the function (now named) `mapOrThrow` throws, while it is `f` which might throw. `mapOrThrow` does the catching.

But I don't insist, it was just a suggestion :wink:

Reply:
Sure, I think it's about different perspectives, each one valid from a different point of view:

There are 2 levels of map: map of `F[_]` and map of `A` to `B`. The user code is responsible for map of `A` to `B`. The function passed to map could be called a `mapper` because it is a mapping from `A` to `B`. So when I see `map`, I interpret that as allowing **me** to map the type parameter to a new type. So `mapOrThrow` means that I may map, or I may throw. Just like `mapOrFail` means that I may map, or I may throw.

Reply:
We used `mapAttempt` for a similar operator as this on `ZIO`.